### ISSUE TITLE
Support of getting the distance maps in DataLoader using multi-workers

### ISSUE THREAD

**Issue by rose-jinyang** [wangchi95/active-boundary-loss#2]
Hello
How are you?
Thanks for contributing to this project.
Your loss works well for right now. But to get the distance maps on GPU is very slow.
So I moved the part into my dataLoader using multi-workers.
However, when running the dataloader, the following issue appears:

![image](https://user-images.githubusercontent.com/47862419/157214375-c4ac79de-c354-404f-a913-79bbb441f8e0.png)

I searched some blogs for such issue and many developers said that the lambda should NOT be used directly in the dataloader using multi-workers.
How can we fix this issue?
Thanks

**Comment by wangchi95** [comment]
I'm not sure where the bug is. But you can refer to the demo code below, which works fine.

```
import torch
from torch.utils.data import Dataset, DataLoader

import numpy as np
from scipy.ndimage import distance_transform_edt as distance
from torchvision import transforms
from functools import partial
from operator import itemgetter

from tqdm import tqdm
import cv2

# Tools
def kl_div(a,b): # q,p
    return F.softmax(b, dim=1) * (F.log_softmax(b, dim=1) - F.log_softmax(a, dim=1))   

def one_hot2dist(seg):
    res = np.zeros_like(seg)
    for i in range(len(seg)):
        posmask = seg[i].astype(np.bool)
        if posmask.any():
            negmask = ~posmask
            res[i] = distance(negmask) * negmask - (distance(posmask) - 1) * posmask
    return res

def class2one_hot(seg, C):
    seg = seg.unsqueeze(dim=0) if len(seg.shape) == 2 else seg
    res = torch.stack([seg == c for c in range(C)], dim=1).type(torch.int32)
    return res

class DummyDataset(Dataset):
        def __init__(self):
            self.dist_map_transform = transforms.Compose([
                lambda img: img.unsqueeze(0),
                lambda nd: nd.type(torch.int64),
                partial(class2one_hot, C=1),
                itemgetter(0),
                lambda t: t.cpu().numpy(),
                one_hot2dist,
                lambda nd: torch.tensor(nd, dtype=torch.float32)
            ])
        
        def get_dist_maps(self, target):
            target_detach = target.clone().detach()
            dist_maps = torch.cat([self.dist_map_transform(target_detach[i]) for i in range(target_detach.shape[0])])
            out = -dist_maps
            out = torch.where(out>0, out, torch.zeros_like(out))
            return out

        def __len__(self):
                return 100

        def __getitem__(self, index):
                dummy_gt_np = np.random.randint(-10,2,[5,20,20]) # random, onehot, [gt_class, h, w]
                dummy_gt_np[dummy_gt_np<0] = 0 # make it sparse
                dummy_gt = torch.from_numpy(dummy_gt_np)
                dist = self.get_dist_maps(dummy_gt)
                return dummy_gt_np, dist

if __name__ == '__main__':
    dummy_dataset = DummyDataset()
    dummy_loader = DataLoader(dummy_dataset, batch_size=2)
    for i, (gt, dist) in enumerate(tqdm(dummy_loader)):
        # cv2.imwrite(f'vis/{i}.png', gt[0][0].numpy().astype(np.uint8)*255)
        # cv2.imwrite(f'vis/{i}_dist.png', dist[0][0].numpy().astype(np.uint8) * 10)
        pass
```

**Comment by rose-jinyang** [comment]
Hello
Thanks for your quick reply.
I mean when using multi-workers for dataloader.

**Comment by wangchi95** [comment]
Oh, I got it. You can replace lambda function with normal function.
like 
```
def tmp_lambda1(x):
    return x.unsqueeze(0)
def tmp_lambda2(x):
    return x.type(torch.int64)
def tmp_lambda3(x):
    return x.cpu().numpy()
def tmp_lambda4(x):
    return torch.tensor(x, dtype=torch.float32)

self.dist_map_transform = transforms.Compose([
    transforms.Lambda(tmp_lambda1),
    transforms.Lambda(tmp_lambda2),
    partial(class2one_hot, C=1),
    itemgetter(0),
    transforms.Lambda(tmp_lambda3),
    one_hot2dist,
    transforms.Lambda(tmp_lambda4),
])
```

**Comment by rose-jinyang** [comment]
Thanks

**Comment by tayton42** [comment]
> Thanks

Hi!Did your method have a significant acceleration effect?Thanks
